internal/gps/vcs_repo.go

```diff
@@ -138,6 +176,25 @@ type hgRepo struct {
 	*vcs.HgRepo
 }
 
+func newHgRepo(ustr, path string) (*hgRepo, error) {
+	r, err := vcs.NewHgRepo(ustr, path)
+	if err != nil {
+		// if vcs could not initialize the repo due to a local error
+		// then the local repo is in an incorrect state. Remove and
+		// treat it as a new not-yet-cloned repo.
+		if _, ok := err.(*vcs.LocalError); ok {
+			os.RemoveAll(path)
+			r, err = vcs.NewHgRepo(ustr, path)
+		}
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &hgRepo{r}, nil
+}
+
```

Comment:
Lots of duplication here... the functions basically vary on `vcs.New*Repo` and their return type. Since the return type is just shoved into a `baseVCSSource` which defines `repo` as a `ctxRepo` interface you can significantly simplify this by making a single function that takes in a ctor, ustr, and path, and returns a ctxRepo (I think).

Rough sketch, not tested at all:

```go
type repoConstructor func(ustr, path string) (ctxRepo, error)

func newRepo(ustr, path string, fn repoConstructor, typeFn repoCreator) (ctxRepo, error) {
  r, err := fn(ustr, path)
  if err != nil {
    // as before ...
  }

  // as before

  return typeFn(r), nil
}
```

```go
type typeFn func(*vcs.Repo) ctxRepo

func newHgRepo(r *vcs.Repo) ctxRepo { return &hgRepo{r} }
// etc
```